File: Singleton.php
<?php

/**
 * Singleton Design Pattern Implementation
 * 
 * The Singleton pattern ensures that a class has only one instance
 * and provides a global point of access to that instance.
 * 
 * Use Cases:
 * - Database connections
 * - Configuration managers
 * - Logging services
 * - Cache managers
 */

class Singleton
{
    /**
     * The single instance of the class
     * @var Singleton|null
     */
    private static ?Singleton $instance = null;

    /**
     * Private constructor to prevent direct instantiation
     */
    private function __construct()
    {
        // Initialize your singleton here
        echo "Singleton instance created!\n";
    }

    /**
     * Prevent cloning of the instance
     */
    private function __clone()
    {
        // Prevent cloning
    }

    /**
     * Prevent unserialization of the instance
     */
    public function __wakeup()
    {
        throw new \Exception("Cannot unserialize singleton");
    }

    /**
     * Get the single instance of the class
     * 
     * @return Singleton
     */
    public static function getInstance(): Singleton
    {
        if (self::$instance === null) {
            self::$instance = new self();
        }
        
        return self::$instance;
    }

    /**
     * Example method
     */
    public function doSomething(): void
    {
        echo "Singleton is doing something!\n";
    }
}

// Example usage:
// $singleton1 = Singleton::getInstance();
// $singleton2 = Singleton::getInstance();
// 
// var_dump($singleton1 === $singleton2); // Output: bool(true)
// $singleton1->doSomething();

?>

File: Singleton.java
/**
 * Singleton Design Pattern Implementation (Java)
 * 
 * Thread-safe implementation using the Bill Pugh Singleton Design
 * This approach uses an inner static helper class to ensure thread safety
 * without requiring synchronization.
 */

public class Singleton {
    
    /**
     * Private constructor to prevent instantiation
     */
    private Singleton() {
        System.out.println("Singleton instance created!");
    }
    
    /**
     * Static inner class - inner classes are not loaded until they are referenced
     * This ensures thread-safe lazy initialization
     */
    private static class SingletonHelper {
        private static final Singleton INSTANCE = new Singleton();
    }
    
    /**
     * Get the single instance of the class
     * 
     * @return Singleton instance
     */
    public static Singleton getInstance() {
        return SingletonHelper.INSTANCE;
    }
    
    /**
     * Example method
     */
    public void doSomething() {
        System.out.println("Singleton is doing something!");
    }
    
    /**
     * Example usage
     */
    public static void main(String[] args) {
        Singleton singleton1 = Singleton.getInstance();
        Singleton singleton2 = Singleton.getInstance();
        
        System.out.println("Are both instances the same? " + (singleton1 == singleton2));
        singleton1.doSomething();
    }
}
